src/Chainweb/Utils.hs

```diff
@@ -367,6 +374,42 @@ data DiffItem a
         -- values.
     deriving (Show, Eq, Ord, Generic, Hashable, Functor, Foldable, Traversable)
 
+mapDiffItem :: (a -> b) -> (a -> b) -> DiffItem a -> DiffItem b
+mapDiffItem f _ (LeftD a) = LeftD (f a)
+mapDiffItem _ g (RightD b) = RightD (g b)
+mapDiffItem f g (BothD a b) = BothD (f a) (g b)
+
+leftDiffItem :: (a -> a) -> DiffItem a -> DiffItem a
+leftDiffItem f = mapDiffItem f id
+{-# INLINE leftDiffItem #-}
+
+rightDiffItem :: (a -> a) -> DiffItem a -> DiffItem a
+rightDiffItem g = mapDiffItem id g
+{-# INLINE rightDiffItem #-}
+
+foldrDiffItem :: (a -> c -> c) -> (a -> c -> c) -> c -> DiffItem a -> c
+foldrDiffItem f _ c (LeftD a) = f a c
+foldrDiffItem _ g c (RightD b) = g b c
+foldrDiffItem f g c (BothD a b) = f a (g b c)
+
+leftItem :: Alternative f => DiffItem a -> f a
+leftItem = foldrDiffItem (const . pure) (\_ _ -> empty) empty
+{-# INLINE leftItem #-}
+
+rightItem :: Alternative f => DiffItem a -> f a
+rightItem = foldrDiffItem (\_ _ -> empty) (const . pure) empty
+{-# INLINE rightItem #-}
+
+traverseDiffItem
+    :: Applicative f
+    => (a -> f b)
+    -> (a -> f b)
+    -> DiffItem a
+    -> f (DiffItem b)
+traverseDiffItem f _ (LeftD a) = LeftD <$> f a
+traverseDiffItem _ g (RightD b) = RightD <$> g b
+traverseDiffItem f g (BothD a b) = BothD <$> f a <*> g b
```

Comment:
This could just be an instance of `Bitraverable`. The above functions could probably be reworked into `Bifoldable` too.

This also looks suspiciously like the `These` datatype.

Reply:
`Bifunctor`, `Bifoldable`, and `Bitraversable` are all parameterized over two types. But `DiffItem` has only a single type parameters. Both elements have the same type.

Reply:
We could consider using something like `type DiffItem a = These a a` (or `Or a a`), which would be a weaker type, but that would still be fine.

What I don't like about `these` is the large set of dependencies. Also that package was a bit unstable recently. `data-or` on the other hand depends only on base but isn't up to date feature wise. So, I think, that a bit of boilerplate here might be fine.

It would be nice if a type like `Or` or `These` existed in base along with the respective `Bi...` instances.